store/gaskvstore.go

```diff
@@ -26,83 +28,88 @@ func NewGasKVStore(gasMeter sdk.GasMeter, gasConfig sdk.GasConfig, parent sdk.KV
+	gs.parent.Set(key, value)
 }
 
 // Implements KVStore.
-func (gi *gasKVStore) Has(key []byte) bool {
-	gi.gasMeter.ConsumeGas(gi.gasConfig.HasCost, "Has")
-	return gi.parent.Has(key)
+func (gs *gasKVStore) Has(key []byte) bool {
+	gs.gasMeter.ConsumeGas(gs.gasConfig.HasCost, sdk.GasHasDesc)
+	return gs.parent.Has(key)
 }
 
 // Implements KVStore.
-func (gi *gasKVStore) Delete(key []byte) {
+func (gs *gasKVStore) Delete(key []byte) {
 	// No gas costs for deletion
-	gi.parent.Delete(key)
+	gs.parent.Delete(key)
 }
 
 // Implements KVStore
-func (gi *gasKVStore) Prefix(prefix []byte) KVStore {
+func (gs *gasKVStore) Prefix(prefix []byte) KVStore {
 	// Keep gasstore layer at the top
 	return &gasKVStore{
-		gasMeter:  gi.gasMeter,
-		gasConfig: gi.gasConfig,
-		parent:    prefixStore{gi.parent, prefix},
+		gasMeter:  gs.gasMeter,
+		gasConfig: gs.gasConfig,
+		parent:    prefixStore{gs.parent, prefix},
 	}
 }
 
 // Implements KVStore
-func (gi *gasKVStore) Gas(meter GasMeter, config GasConfig) KVStore {
-	return NewGasKVStore(meter, config, gi)
+func (gs *gasKVStore) Gas(meter GasMeter, config GasConfig) KVStore {
+	return NewGasKVStore(meter, config, gs)
 }
 
-// Implements KVStore.
-func (gi *gasKVStore) Iterator(start, end []byte) sdk.Iterator {
-	return gi.iterator(start, end, true)
+// Iterator implements the KVStore interface. It returns an iterator which
+// incurs a flat gas cost of iterator initialization.
+func (gs *gasKVStore) Iterator(start, end []byte) sdk.Iterator {
+	return gs.iterator(start, end, true)
 }
 
-// Implements KVStore.
-func (gi *gasKVStore) ReverseIterator(start, end []byte) sdk.Iterator {
-	return gi.iterator(start, end, false)
+// ReverseIterator implements the KVStore interface. It returns a reverse
+// iterator which incurs a flat gas cost of iterator initialization.
+func (gs *gasKVStore) ReverseIterator(start, end []byte) sdk.Iterator {
+	return gs.iterator(start, end, false)
 }
 
 // Implements KVStore.
-func (gi *gasKVStore) CacheWrap() sdk.CacheWrap {
+func (gs *gasKVStore) CacheWrap() sdk.CacheWrap {
 	panic("cannot CacheWrap a GasKVStore")
 }
 
 // CacheWrapWithTrace implements the KVStore interface.
-func (gi *gasKVStore) CacheWrapWithTrace(_ io.Writer, _ TraceContext) CacheWrap {
+func (gs *gasKVStore) CacheWrapWithTrace(_ io.Writer, _ TraceContext) CacheWrap {
 	panic("cannot CacheWrapWithTrace a GasKVStore")
 }
 
-func (gi *gasKVStore) iterator(start, end []byte, ascending bool) sdk.Iterator {
+func (gs *gasKVStore) iterator(start, end []byte, ascending bool) sdk.Iterator {
 	var parent sdk.Iterator
 	if ascending {
-		parent = gi.parent.Iterator(start, end)
+		parent = gs.parent.Iterator(start, end)
 	} else {
-		parent = gi.parent.ReverseIterator(start, end)
+		parent = gs.parent.ReverseIterator(start, end)
 	}
-	return newGasIterator(gi.gasMeter, gi.gasConfig, parent)
+
+	gs.gasMeter.ConsumeGas(gs.gasConfig.IterInitCostFlat, sdk.GasIterInitCostFlatDesc)
```

Comment:
I think when we call `parent.Iterator` (or `parent.ReverseIterator`) it already seeks to the first value, so we should charge key and value cost as with `.Next()`. I don't know if we need a flat cost but having that too is definitely safe.

Reply:
Will amend to have initialization incur a cost of akin to `.Next()` -- this will actually result in the removal of the `IterInitCostFlat`.

Reply:
also you have to double-check `Valid()` here and maybe below too before calling `.Value()`

Reply:
otherwise it will panic!

Reply:
Indeed.